### PR TITLE
Accelerate builds in VS by copying files directly

### LINKED ISSUE #4665: Consuming projects are not updated when a CopyToOutputDirectory item is updated in a project with no project references.
**Visual Studio Version**: 2017 - 15.9.10

Also reproduced in 2019 RC.4 - 16.0.0

**Summary:**

Fast up-to-date checks will consider a project to be up-to-date even when a project it references is not up to date due to a stale `CopyToOutputDirectory` item.

(I've found some workarounds which are [described here](https://github.com/PathogenPlayground/TransitiveCopiedConentNotUpdated#workarounds).)

**Steps to Reproduce**:

1. [Clone this repository](https://github.com/PathogenPlayground/TransitiveCopiedConentNotUpdated).
    * The demonstration repository contains details on how the projects are set up. The most important thing to know is that `TextFileA.txt` and `TextFileB.txt` are copied to the output directory with `PreserveNewest` and the program prints their contents when it runs.
    * The repository also contains more details on my findings about the source of the issue.
    * The steps from the repository for reproducing the problem are reproduced below for posterity.
1. Open the solution in Visual Studio.
2. Select `TestProgram` as the startup project.
3. Press F5 to start the program. The output is as follows:
    ```
    Hello World!
    Hello from ClassA!
    Text file A reporting for duty!
    Hello from ClassB!
    Text file B reporting for duty!
    Done.
    ```
4. Close the program and add a message `TextFileA.txt` and press F5 to start the program again. The output includes your message:
    ```
    Hello World!
    Hello from ClassA!
    Text file A reporting for duty! Hello, world!
    Hello from ClassB!
    Text file B reporting for duty!
    Done.
    ```
5. Close the program and add a message to `TextFileB.txt` instead, press F5 to start the program again. **The output will not include your message:**
    ```
    Hello World!
    Hello from ClassA!
    Text file A reporting for duty! Hello, world!
    Hello from ClassB!
    Text file B reporting for duty!
    Done.
    ```
6. Observe that the file is updated in `LibraryB`'s output (`LibraryB\bin\Debug\netstandard2.0\TextFileB.txt`), but not in `TestProgram`'s output (`TestProgram\bin\Debug\netcoreapp2.2\TextFileB.txt`).

**Expected Behavior**:

`TextFileB.txt` is updated in `TestProgram`'s output folder when it is changed.

**Actual Behavior**:

`TextFileB.txt` is not updated.

**User Impact**:

Ironically I actually found this bug while debugging a similar issue with a legacy-style C# project. Either way, the symptoms are the same even though it appears the cause is not.

This issue is very confusing because you can see that `LibraryB` was built by Visual Studiom, but projects depending on it directly/indirectly were not. The copied file will be present in the output of `LibraryB`, but not in `TestProgram`. You can rebuild the solution or make dummy changes in `TestProgram` to force things to update, but this is fairly annoying to remember to do. In our case, the copied file was a native library, so it was like changes to the native library weren't working.

**Workarounds**:

Besides rebuilding, I've found two workarounds to this issue:
1. Add a dummy reference to the library.
2. Create the copy marker yourself using MSBuild.

Both are [described in further detail in my demonstration repository](https://github.com/PathogenPlayground/TransitiveCopiedConentNotUpdated#workarounds).

### PR DESCRIPTION
**Pull request by drewnoakes**
Fixes #4665

This adds an opt-in feature to have Visual Studio copy files directly during build, bypassing MSBuild when it appears safe to do so.

This can reduce build times considerably, especially in solutions with large project reference graphs.

See the included `docs/build-acceleration.md` file for details on the feature.

This change also increases the FUTDC's visibility on inputs from other projects, meaning the underbuild issue reported in #4665 is addressed as part of this change.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/8696)

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs [hunk 1]
```diff
@@ -605,115 +672,121 @@ private bool CheckMarkers(Log log, in TimestampCache timestampCache, UpToDateChe
...
             return true;
         }
 
-        private bool CheckCopyToOutputDirectoryFiles(Log log, in TimestampCache timestampCache, UpToDateCheckImplicitConfiguredInput state, CancellationToken token)
+        private bool CheckCopyToOutputDirectoryItems(Log log, in TimestampCache timestampCache, UpToDateCheckImplicitConfiguredInput state, IEnumerable<CopyItem> copyItems, ConfiguredFileSystemOperationAggregator fileSystemAggregator, CancellationToken token)
         {
             string outputFullPath = Path.Combine(state.MSBuildProjectDirectory, state.OutputRelativeOrFullPath);
 
-            foreach ((string itemType, ImmutableArray<UpToDateCheckInputItem> items) in state.InputSourceItemsByItemType)
+            bool isLogging = log.Level >= LogLevel.Verbose && state.CopyToOutputDirectoryItems.Length != 0;
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs:732]: The `state.CopyToOutputDirectoryItems.Length` check looks wrong.

**Review by melytc — APPROVED** [review]
All looks good to me! Appreciated the documentation on the comments 🙂

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: Fixed in 73f948f50c69f0a68257e30fc0746d17b26733cb.

**Review by MiYanni — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.CopyType.cs [hunk 2]
```diff
@@ -4,9 +4,11 @@ namespace Microsoft.VisualStudio.ProjectSystem.UpToDate
 {
     internal sealed partial class BuildUpToDateCheck
     {
+        /// <summary>
+        /// Potential values for the <c>CopyToOutputDirectory</c> metadata on items.
+        /// </summary>
         internal enum CopyType
         {
-            Never,
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.CopyType.cs:9]: We used to model the project's inputs using a type called `UpToDateCheckInputItem` which modelled the copy type. Not all of these items were copied, so we needed a `Never` value. 

With the changes here, we model inputs and copy items separately. Inputs become just a `string` file path, and the `UpToDateCheckInputItem` type is removed. The new `CopyItem` type is only used for copyable files, so we no longer need the `Never` value.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/CopyItemAggregator.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        }
+    }
+
+    public (IEnumerable<CopyItem> Items, bool IsComplete) TryGatherCopyItemsForProject(string targetPath, BuildUpToDateCheck.Log logger)
+    {
+        // Keep track of all projects we've visited to avoid infinite recursion or duplicated results.
+        HashSet<string> explored = new(StringComparers.Paths);
+
+        // The queue of projects yet to be visited.
+        Queue<string> frontier = new();
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/CopyItemAggregator.cs:41]: This algorithm is walking the project reference graph, starting at the project being checked, and gathering all reachable copy items.

We use a queue to walk items. The _frontier_ is the outward facing edge of that graph exploration, and tracks the copy items at each project.

It's a name I've seen used in academia: https://www.cs.dartmouth.edu/~scot/cs10/lectures/19/19.html

[the remark this answers is not in the document]
Comment (MiYanni) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/CopyItemAggregator.cs:41]: Interesting.

**Pull request merged**